pilot/pkg/xds/ads.go

```diff
@@ -775,22 +773,24 @@ func (s *DiscoveryServer) pushConnection(con *Connection, pushEv *Event) error {
 		log.Debugf("Skipping push to %v, no updates required", con.conID)
 		if pushRequest.Full {
 			// Only report for full versions, incremental pushes do not have a new version.
-			reportAllEvents(s.StatusReporter, con.conID, pushRequest.Push.LedgerVersion, nil)
+			reportAllEvents(con, s.StatusReporter, pushRequest.Push.LedgerVersion)
 		}
 		return nil
 	}
 
 	// Send pushes to all generators
 	// Each Generator is responsible for determining if the push event requires a push
-	wrl, ignoreEvents := con.pushDetails()
+	wrl := con.orderWatchedResources()
 	for _, w := range wrl {
 		if err := s.pushXds(con, w, pushRequest); err != nil {
 			return err
 		}
 	}
+
 	if pushRequest.Full {
-		// Report all events for unwatched resources. Watched resources will be reported in pushXds or on ack.
-		reportAllEvents(s.StatusReporter, con.conID, pushRequest.Push.LedgerVersion, ignoreEvents)
+		// This reports all the tracking types event, including that are not watched.
+		// It does not mean the xds response is received and loaded by envoy.
```

Comment:
It SHOULD for types we push.. just not for types we did not push.

Reply:
Including types we donot push for gateway like in some initial state, we donot need to push rds to a gateway. But we need to record that for rds, otherwise wait will be stuck.

And i have tested this comemnt is true
https://github.com/istio/istio/pull/48542#pullrequestreview-1800732708